itest/lnd_channel_funding_fund_max_test.go

```diff
@@ -0,0 +1,277 @@
+				)
+			},
+		)
+
+		// Stop at the first failure. Mimic behavior of original test
+		// framework.
+		if !success {
+			break
+		}
+	}
+}
+
+// runTestCase runs a single test case asserting that test conditions are met.
+func runFundMaxTestCase(ht *lntest.HarnessTest, t *testing.T, alice,
+	bob *node.HarnessNode, testCase *chanFundMaxTestCase) {
+
+	ht.FundCoins(testCase.initialWalletBalance, alice)
+
+	defer func() {
+		if testCase.initialWalletBalance <= 2_000 {
+			// Add additional funds to sweep "dust" UTXO.
+			ht.FundCoins(100_000, alice)
+		}
+
+		// Remove all funds from Alice.
+		sweepNodeWalletAndAssert(ht, alice)
+	}()
+
+	// The parameters to try opening the channel with.
+	chanParams := lntest.OpenChannelParams{
+		Amt:         0,
+		PushAmt:     testCase.pushAmt,
+		SatPerVByte: testCase.feeRate,
+		FundMax:     true,
+		Private:     testCase.private,
+	}
+
+	// If we don't expect the channel opening to be
+	// successful, simply check for an error.
+	if testCase.chanOpenShouldFail {
+
+		expectedErr := fmt.Errorf(testCase.expectedErrStr)
+		ht.OpenChannelAssertErr(
+			alice, bob, chanParams, expectedErr,
+		)
+
+		return
+	}
+
+	// Otherwise, if we expect to open a channel use the helper function.
+	chanPoint := ht.OpenChannel(alice, bob, chanParams)
+
+	// Close the channel between Alice and Bob, asserting
+	// that the channel has been properly closed on-chain.
+	defer ht.CloseChannel(alice, chanPoint)
+
+	cType := ht.GetChannelCommitType(alice, chanPoint)
+
+	// Alice's balance should be her amount subtracted by the commitment
+	// transaction fee.
+	checkChannelBalance(
+		ht, alice,
+		testCase.expectedBalanceAlice-lntest.CalcStaticFee(cType, 0),
+		testCase.pushAmt,
+	)
+
+	// Ensure Bob's balance within the channel is equal to the push amount.
+	checkChannelBalance(
+		ht, bob, testCase.pushAmt,
+		testCase.expectedBalanceAlice-lntest.CalcStaticFee(cType, 0),
+	)
+}
+
+// Creates a helper closure to be used below which asserts the proper
+// response to a channel balance RPC.
+func checkChannelBalance(ht *lntest.HarnessTest, node *node.HarnessNode,
+	local, remote btcutil.Amount) {
+
+	expectedResponse := &lnrpc.ChannelBalanceResponse{
+		LocalBalance: &lnrpc.Amount{
+			Sat:  uint64(local),
+			Msat: uint64(lnwire.NewMSatFromSatoshis(local)),
+		},
+		RemoteBalance: &lnrpc.Amount{
+			Sat: uint64(remote),
+			Msat: uint64(lnwire.NewMSatFromSatoshis(
+				remote,
+			)),
+		},
+		UnsettledLocalBalance:    &lnrpc.Amount{},
+		UnsettledRemoteBalance:   &lnrpc.Amount{},
+		PendingOpenLocalBalance:  &lnrpc.Amount{},
+		PendingOpenRemoteBalance: &lnrpc.Amount{},
+		// Deprecated fields.
+		Balance: int64(local),
+	}
+	ht.AssertChannelBalanceResp(node, expectedResponse)
+}
+
+// fundingFee returns the fee estimate used for a tx with the given number of
+// inputs and the optional change output. This matches the estimate done by the
+// wallet.
+func fundingFee(numInput int, change bool) btcutil.Amount {
```

Comment:
Q: Is it common in itests to replicate the functions in the codebase rather then use them here?

Reply:
I doubt that this is a common pattern.

Reply:
is it better to replicate the code rather than using the feecalculation form the actual code? Asking out of curiosity how to do this testing, my suggestion would come with the risk of testing way too "whitetesty" but depending on why you wrote this function you basically looked how the real source is doing it an replicated it so not sure whats the best way to test in such circumstances ?

Reply:
Yes good point. I think (initially this function)[https://github.com/lightningnetwork/lnd/pull/4029/files#diff-583407cd98e4e423efd8a6afe4acb5b67395da6ae79927b11f9e3b989f822313R71] was added for readability and brevity as it consumes a `P2WKH` and ouputs a `P2WSH` plus change whereas `calculateFees` does fee calc in a comprehensive way.
But I agree that we should use the actual fee calc done in the backend so we are not missing future changes there.
I will change the test fee calc to use the backend calc.